mlos_bench/mlos_bench/run.py

```diff
@@ -154,29 +123,96 @@ def _optimize(*,
     return (best_score, best_config)
 
 
-def _run(env: Environment, opt: Optimizer, trial: Storage.Trial, global_config: Dict[str, Any]) -> None:
+def _load_config(exp: Storage.Experiment, env_context: Environment,
+                 config_id: int) -> TunableGroups:
+    """
+    Load the existing tunable configuration from the storage.
+    """
+    tunable_values = exp.load_tunable_config(config_id)
+    tunables = env_context.tunable_params.assign(tunable_values)
+    _LOG.info("Load config from storage: %d", config_id)
+    if _LOG.isEnabledFor(logging.DEBUG):
+        _LOG.debug("Config %d ::\n%s",
+                   config_id, json.dumps(tunable_values, indent=2))
+    return tunables
+
+
+def _scheduler(exp: Storage.Experiment, env_context: Environment,
+               global_config: Dict[str, Any], running: bool = False) -> None:
+    """
+    Scheduler part of the loop. Check for pending trials in the queue and run them.
+    """
+    for trial in exp.pending_trials(datetime.utcnow(), running=running):
+        _run_trial(env_context, trial, global_config)
+
+
+def _optimizer(exp: Storage.Experiment, opt_context: Optimizer,
```

Comment:
```suggestion
def _get_optimizer_suggestions(exp: Storage.Experiment, opt_context: Optimizer,
```

Am thinking that actual scheduling, should be left up to the Scheduler class, and the Optimizer should only be in charge of suggesting new configs.

Now, especially when it comes to multi-objective and/or parallel trials, it might be worthwhile for there to be some sort of feedback between the two that allows, for instance, the Scheduler to accept additional information like cost/budget/uncertainty/priority/weight in from the Optimizer, but I think actually determining which of the subset of options the Optimizer has presented should be run next, should be part of the Scheduler.

Reply:
Also, if the Scheduler is swappable, it's possible that the config changes mid run, and a new Scheduler policy gets used after some other ones were already enqueued to the pending trials queue.
Not entirely certain how those might interact.  Maybe scheduling only happens on entry to the queue, and on popping items off the queue at any given iteration, we can reorder with the current policy if there are any ties (e.g., equal weights and start times or some such), but otherwise not have much smart logic for that.  i.e., it should mostly be admission control policies.

Thoughts?

Reply:
yes, I was not exactly happy with the naming of my functions, either 🙂 - thanks for the suggestions!

My plan is to implement the swappable Scheduler class in the next PR. The purpose of this one is to disentangle the trial execution and config recommendation parts and see what other components have to be updated to accommodate this change. This way, we keep the PRs small, and the next PR won't have updates to the storage, optimizer, etc.